llmfoundry/utils/exceptions.py

```diff
@@ -0,0 +1,215 @@
+# Copyright 2024 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Custom exceptions for the LLMFoundry."""
```

Comment:
I thought about doing more general errors, something like this:
```
class FinetuneDataFormatError(Exception):
    def __init__(self, message: str, attributes: List[Any], errorType: enum) -> None:
        self.finetuneErrorType = errorType
        self.attributes = attributes
        super().__init__(message)

raise FinetuneDataFormatError(f'Expected 2 keys in message, but found {len(keys)}', [keys], IncorrectMessageKeyQuantityError)
```
so that in mcloud, we would get the finetuneErrorType and attributes and map it to "Expected 2 keys in each chat message (role and content), but found keys ${attributes.keys} in at least one message in your train dataset."

I chose against this because I found that supporting the message we want to raise in foundry and the required data we need for the mcloud user message construction is harder to maintain if it is distributed in multiple files throughout the codebase 

and we don't want to hard code the user message in foundry because it's awk to show those open source and it's harder to make changes for quick release

Reply:
sgtm!